File: orig_app/internal/handler/storage/static_object_external_storage.go
package storage

import (
	"crypto/subtle"
	"net/http"

	"gitlab.com/gitlab-org/gitlab-redux/internal/handler/application"
	"gitlab.com/gitlab-org/gitlab-redux/internal/model"
	"gitlab.com/gitlab-org/gitlab-redux/internal/utils/errors"
)

// StaticObjectExternalStorage provides functionality for handling external storage
// for static objects
type StaticObjectExternalStorage struct {
	settingsService    *model.SettingsService
	applicationHelper *application.Helper
}

// NewStaticObjectExternalStorage creates a new instance of StaticObjectExternalStorage
func NewStaticObjectExternalStorage(
	settingsService *model.SettingsService,
	applicationHelper *application.Helper,
) *StaticObjectExternalStorage {
	return &StaticObjectExternalStorage{
		settingsService:    settingsService,
		applicationHelper: applicationHelper,
	}
}

// RedirectToExternalStorage redirects to external storage if the request is not
// already from external storage
func (s *StaticObjectExternalStorage) RedirectToExternalStorage(w http.ResponseWriter, r *http.Request, project *model.Project) error {
	if s.IsExternalStorageRequest(r) {
		return nil
	}

	externalURL := s.applicationHelper.ExternalStorageURLOrPath(r.URL.Path, project)
	http.Redirect(w, r, externalURL, http.StatusTemporaryRedirect)
	return nil
}

// IsExternalStorageRequest checks if the request is from external storage by validating
// the X-Gitlab-External-Storage-Token header
func (s *StaticObjectExternalStorage) IsExternalStorageRequest(r *http.Request) bool {
	headerToken := r.Header.Get("X-Gitlab-External-Storage-Token")
	if headerToken == "" {
		return false
	}

	externalStorageToken := s.settingsService.StaticObjectsExternalStorageAuthToken()
	if externalStorageToken == "" {
		return false
	}

	// Use constant-time comparison to prevent timing attacks
	if subtle.ConstantTimeCompare([]byte(headerToken), []byte(externalStorageToken)) == 1 {
		return true
	}

	// If token is present but invalid, raise access denied error
	panic(errors.NewAccessDeniedError("invalid external storage token"))
}

File: rust_app/src/controllers/concerns/static_object_external_storage.rs
// Ported from: orig_app/app/controllers/concerns/static_object_external_storage.rb
//
// This trait provides logic for redirecting to external storage and authenticating requests
// using an external storage token. Intended for use in Actix-web controllers.

use crate::config::settings::Settings;
use actix_web::{http::header, web, HttpRequest, HttpResponse};
use subtle::ConstantTimeEq;

pub trait StaticObjectExternalStorage {
    /// Checks if the request is from external storage by validating the token header.
    fn external_storage_request(&self, req: &HttpRequest) -> bool {
        let header_token = req.headers().get("X-Gitlab-External-Storage-Token");
        let header_token = match header_token.and_then(|h| h.to_str().ok()) {
            Some(token) if !token.is_empty() => token,
            _ => return false,
        };
        let settings = Settings::current();
        let external_storage_token = settings.static_objects_external_storage_auth_token.as_str();
        // Use constant-time comparison for security
        if header_token
            .as_bytes()
            .ct_eq(external_storage_token.as_bytes())
            .into()
        {
            true
        } else {
            // In Ruby: raise(Gitlab::Access::AccessDeniedError)
            // In Rust: could return false or handle error as needed
            false
        }
    }

    /// Redirects to external storage if the request is not already from external storage.
    fn redirect_to_external_storage(
        &self,
        req: &HttpRequest,
        project: &str,
    ) -> Option<HttpResponse> {
        if self.external_storage_request(req) {
            return None;
        }
        // Use the existing concern for URL generation
        let fullpath = req
            .uri()
            .path_and_query()
            .map(|pq| pq.as_str())
            .unwrap_or("");
        let url = crate::controllers::concerns::static_object_external_storage_url::StaticObjectExternalStorageUrl::get_external_storage_url(self, fullpath);
        if let Some(external_url) = url {
            Some(
                HttpResponse::Found()
                    .append_header((header::LOCATION, external_url))
                    .finish(),
            )
        } else {
            None
        }
    }
}
